histolab/tile.py

```diff
@@ -166,6 +166,53 @@ def save(self, path: Union[str, bytes, os.PathLike]) -> None:
         Path(path).parent.mkdir(parents=True, exist_ok=True)
         self._image.save(path)
 
+    @lazyproperty
+    def tissue_mask(self) -> np.ndarray:
+        """Return a binary mask representing the tissue in the tile.
+
+        Returns
+        -------
+        np.ndarray
+            Binary mask representing the tissue in the tile.
+
+        Notes
+        -----
+        In order to address an issue with the calculation of the tissue mask in tiles
+        where the tissue covers all the tile area, the filters are applied on the tile
+        with a white border of 10px, and then the border is removed from the returned
+        mask.
+        """
+        np_tile = np.array(self._image)
+
+        border_thickness = 10
+
+        # create a "white" array 20px bigger than the tile
+        container_size = (
+            np_tile.shape[0] + 2 * border_thickness,
+            np_tile.shape[1] + 2 * border_thickness,
+            np_tile.shape[2],
+        )
+        container = np.ones(container_size, dtype=np.uint8) * 255
+
+        # paste the tile at the center
+        container[
+            border_thickness : np_tile.shape[0] + border_thickness,
+            border_thickness : np_tile.shape[1] + border_thickness,
+            :,
+        ] = np_tile
+        tile_border = PIL.Image.fromarray(container)
+
+        # apply filters on tile with border
+        filters = FiltersComposition(Tile).tissue_mask_filters
+        mask_border = filters(tile_border)
+
+        # remove border
+        mask = mask_border[
+            border_thickness : np_tile.shape[0] + border_thickness,
+            border_thickness : np_tile.shape[1] + border_thickness,
+        ]
+        return mask
```

Comment:
General concern of this property is the massive usage of variables. Considering that this is a lazyproperty and it will be cached we are using a lot of memory for storing memory.
For example:
`tile_border = PIL.Image.fromarray(container)` is really needed?
IMHO we should try to optimize the memory of this property in order to keep it as small as possible.

Reply:
`lru_cache` works in a way that it saves in a dictionary **only** the result of the called function, so honestly I wouldn't be concerned with the amount of variables we use internally (I can agree that it could be "heavy" the first time).

That considered, it is a "big" function, doing a task that I don't know how to reasonably break into sub-tasks, and my main focus when writing it was that it would be easy to read and understand. What do you think @ernestoarbitrio ?

Reply:
I'm not saying I would split into little pieces. I'm just saying to check if we can get rid of some assigment to free-up memory. Think of a pipeline where histolab is used. If this property is called in a foorloop or basically is called 1000 times we have to be sure we are doing memory de-allocation. What do you think?

Reply:
To answer some of your questions in a data driven-way, I rescued to the use of a memory profiler.

**This is the implementation with reduced variables assignments**:
```python
    @lazyproperty
    def tissue_mask_less_assignments(self) -> np.ndarray:
        """Binary mask representing the tissue in the tile.

        Returns
        -------
        np.ndarray
            Binary mask representing the tissue in the tile.

        Notes
        -----
        In order to address an issue with the calculation of the tissue mask in tiles
        where the tissue covers all the tile area, the filters are applied on the tile
        with a white border of 10px, and then the border is removed from the returned
        mask.
        """
        np_tile = np.array(self._image)

        border_thickness = 10

        # create a "white" array 20px bigger than the tile
        container_size = ()
        container = (
            np.ones(
                (
                    np_tile.shape[0] + 2 * border_thickness,
                    np_tile.shape[1] + 2 * border_thickness,
                    np_tile.shape[2],
                ),
                dtype=np.uint8,
            )
            * 255
        )

        # paste the tile at the center
        container[
            border_thickness : np_tile.shape[0] + border_thickness,
            border_thickness : np_tile.shape[1] + border_thickness,
            :,
        ] = np_tile

        # apply filters on tile with border
        filters = FiltersComposition(Tile).tissue_mask_filters
        mask_border = filters(PIL.Image.fromarray(container))

        # remove border
        return mask_border[
            border_thickness : np_tile.shape[0] + border_thickness,
            border_thickness : np_tile.shape[1] + border_thickness,
        ]
```

**This is the code I used to profile the current implementation**:
```python
from memory_profiler import profile

from histolab.tile import Tile
from PIL import Image

img = Image.open("tests/fixtures/tiles/high-nuclei-score-level2.png")

tile = Tile(img, None, None)


@profile
def normal():
    a = tile.tissue_mask
    b = tile.tissue_mask
    c = tile.tissue_mask
    d = tile.tissue_mask
    e = tile.tissue_mask
    f = tile.tissue_mask
    g = tile.tissue_mask


normal()
```

**And this is the code I used to profile the modified implementation**:
```python
from memory_profiler import profile

from histolab.tile import Tile
from PIL import Image

img = Image.open("tests/fixtures/tiles/high-nuclei-score-level2.png")

tile = Tile(img, None, None)


@profile
def less_assignments():
    a = tile.tissue_mask_less_assignments
    b = tile.tissue_mask_less_assignments
    c = tile.tissue_mask_less_assignments
    d = tile.tissue_mask_less_assignments
    e = tile.tissue_mask_less_assignments
    f = tile.tissue_mask_less_assignments
    g = tile.tissue_mask_less_assignments


less_assignments()
```
## Results with 512x512 tile (140KB):

**Current**:
```Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    11     98.4 MiB     98.4 MiB           1   @profile
    12                                         def normal():
    13    106.9 MiB      8.6 MiB           1       a = tile.tissue_mask
    14    106.9 MiB      0.0 MiB           1       b = tile.tissue_mask
    15    106.9 MiB      0.0 MiB           1       c = tile.tissue_mask
    16    106.9 MiB      0.0 MiB           1       d = tile.tissue_mask
    17    106.9 MiB      0.0 MiB           1       e = tile.tissue_mask
    18    106.9 MiB      0.0 MiB           1       f = tile.tissue_mask
    19    106.9 MiB      0.0 MiB           1       g = tile.tissue_mask
```

**Modified**:
```
Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    11     98.4 MiB     98.4 MiB           1   @profile
    12                                         def less_assignments():
    13    107.0 MiB      8.6 MiB           1       a = tile.tissue_mask_less_assignments
    14    107.0 MiB      0.0 MiB           1       b = tile.tissue_mask_less_assignments
    15    107.0 MiB      0.0 MiB           1       c = tile.tissue_mask_less_assignments
    16    107.0 MiB      0.0 MiB           1       d = tile.tissue_mask_less_assignments
    17    107.0 MiB      0.0 MiB           1       e = tile.tissue_mask_less_assignments
    18    107.0 MiB      0.0 MiB           1       f = tile.tissue_mask_less_assignments
    19    107.0 MiB      0.0 MiB           1       g = tile.tissue_mask_less_assignments
```

## Results with 956 × 779 tile (1.7MB):

**Current**:
```
Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    11     98.5 MiB     98.5 MiB           1   @profile
    12                                         def normal():
    13    114.6 MiB     16.1 MiB           1       a = tile.tissue_mask
    14    114.6 MiB      0.0 MiB           1       b = tile.tissue_mask
    15    114.6 MiB      0.0 MiB           1       c = tile.tissue_mask
    16    114.6 MiB      0.0 MiB           1       d = tile.tissue_mask
    17    114.6 MiB      0.0 MiB           1       e = tile.tissue_mask
    18    114.6 MiB      0.0 MiB           1       f = tile.tissue_mask
    19    114.6 MiB      0.0 MiB           1       g = tile.tissue_mask
```

**Modified**:
```
Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    11     98.4 MiB     98.4 MiB           1   @profile
    12                                         def less_assignments():
    13    114.5 MiB     16.1 MiB           1       a = tile.tissue_mask_less_assignments
    14    114.5 MiB      0.0 MiB           1       b = tile.tissue_mask_less_assignments
    15    114.5 MiB      0.0 MiB           1       c = tile.tissue_mask_less_assignments
    16    114.5 MiB      0.0 MiB           1       d = tile.tissue_mask_less_assignments
    17    114.5 MiB      0.0 MiB           1       e = tile.tissue_mask_less_assignments
    18    114.5 MiB      0.0 MiB           1       f = tile.tissue_mask_less_assignments
    19    114.5 MiB      0.0 MiB           1       g = tile.tissue_mask_less_assignments
```

## Conclusions
The important bits of the profiler results are the increments on the first line/first call. They are exactly the same between the two implementations with a smaller and a bigger tile. Therefore I don't see the point in using fewer variables (which would just damage the readability)

Reply:
awesome great answer